### PR TITLE
Un-balanced logging properly supported

### LINKED ISSUE #5063: Self.log with multiple Optimizers errors out in 1.1.0 but works in 1.0.8
## 🐛 Bug

self.log results in the following error:
```
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py", line 207, in auto_reduce_results_on_epoch_end
    opt_outputs = epoch_metrics[opt_idx]
```
when the following conditions are satisfied:

- There are two optimizers
- 1 of the optimizers does not log anything in 1 training epoch

Note, that this bug is not present in v1.0.8 and was introduced in 1.1.0

## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

https://colab.research.google.com/drive/1MkMmuzTmZU2hkPjoEQ8Mwd4qSh168QKn?usp=sharing

### To Reproduce
See notebook

### Expected behavior
The same behavious as 1.0.8.

### Environment
See notebook.

### Additional context
See notebook.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolves an indexation typo and simplify slightly the code.

Fixes #5063

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/logging_tests/test_train_loop_logging_1_0.py [hunk 1]
```diff
@@ -771,3 +771,34 @@ def on_train_epoch_end(self, *_):
     trainer.fit(model)
     assert model.epoch_end_called
     assert model.on_train_epoch_end_called
+
+
+def test_unbalanced_logging_with_multiple_optimizers(tmpdir):
```
Comment (Borda) [inline tests/trainer/logging_tests/test_train_loop_logging_1_0.py:776]: shall we rather have a special test/file for multi-optimizer cases?

**Review by Borda — APPROVED** [review]
lgtm

File: tests/trainer/logging_tests/test_train_loop_logging_1_0.py (hunk 1 above)
Comment (tchaton) [reply tests/trainer/logging_tests/test_train_loop_logging_1_0.py:776]: Why not, let me make the change.

**Review by s-rog — APPROVED** [review]

File: tests/trainer/optimization/test_multiple_optimizers.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+            optimizer2 = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+            return [optimizer, optimizer2]
+
+    model = TestModel()
+    model.training_epoch_end = None
+
+    # Initialize a trainer
+    trainer = pl.Trainer(
+        max_epochs=2,
```
Comment (rohitgr7) [inline tests/trainer/optimization/test_multiple_optimizers.py:48]:
```suggestion
        default_root_dir=tmpdir,
        max_epochs=2,
```

**Review by rohitgr7 — APPROVED** [review]

File: tests/trainer/optimization/test_multiple_optimizers.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    model = TestModel()
+    model.training_epoch_end = None
+
+    # Initialize a trainer
+    trainer = pl.Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+    )
+
+    trainer.fit(model)
```
Comment (carmocca) [inline tests/trainer/optimization/test_multiple_optimizers.py:52]: Does this test fail on master? Can you add some asserts to check the logging state?

Comment (Borda) [reply tests/trainer/optimization/test_multiple_optimizers.py:52]: good point, pls add it...

Comment (tchaton) [reply tests/trainer/optimization/test_multiple_optimizers.py:57]: Yes, thanks.

**Review by justusschock — APPROVED** [review]
